plugin/src/main/java/io/jenkins/plugins/casc/impl/configurators/HeteroDescribableConfigurator.java

```diff
@@ -57,117 +64,168 @@ public HeteroDescribableConfigurator(Class<T> clazz) {
+    @SuppressWarnings("unchecked")
+    private Class<T> descriptorClass(Descriptor<T> descriptor) {
+        return descriptor.getKlass().toJavaClass();
+    }
 
-        final Class api = getImplementedAPI();
-        // Search for @Symbol annotation on Descriptor to match shortName
+    private Option<Descriptor<T>> lookupDescriptor(String symbol, CNode config) {
+        return getDescriptors()
+                .filter(descriptor -> findByPreferredSymbol(descriptor, symbol) || findBySymbols(descriptor, symbol, config))
+                .map(descriptor -> Tuple.of(preferredSymbol(descriptor), descriptor))
+                .foldLeft(HashMap.empty(), this::handleDuplicateSymbols)
+                .values()
+                .headOption()
+                .orElse(() -> {
+                    String message = "No " + target.getName() + " implementation found for " + symbol;
+                    String possible = lookupPlugin("configuration-as-code-support") ? "" : System.lineSeparator() + "Possible solution: Try to install 'configuration-as-code-support' plugin";
+                    throw new IllegalArgumentException(Stream.of(message, possible)
+                            .intersperse("")
+                            .foldLeft(new StringBuilder(), StringBuilder::append)
+                            .toString());
+                });
+    }
 
-        for (Descriptor d : candidates) {
-            final List<String> symbols = DescribableAttribute.getSymbols(d, api, target);
-            final String preferred = symbols.get(0);
-            if (preferred.equalsIgnoreCase(shortname)) {
-                return d.getKlass().toJavaClass();
-            }
+    private Boolean findByPreferredSymbol(Descriptor<T> descriptor, String symbol) {
+        return preferredSymbol(descriptor).equalsIgnoreCase(symbol);
+    }
+
+    private Boolean findBySymbols(Descriptor<T> descriptor, String symbol, CNode node) {
+        return getSymbols(descriptor)
+                .find(actual -> actual.equalsIgnoreCase(symbol))
+                .map(actual -> {
+                    ObsoleteConfigurationMonitor.get().record(node, "'" + symbol + "' is obsolete, please use '" + preferredSymbol(descriptor) + "'");
+                    return descriptorClass(descriptor);
+                }).isDefined();
+    }
+
+    private Stream<String> getSymbols(Descriptor<T> descriptor) {
+        return Stream.ofAll(DescribableAttribute.getSymbols(descriptor, target, target));
+    }
+
+    private String preferredSymbol(Descriptor<?> descriptor) {
+        return DescribableAttribute.getPreferredSymbol(descriptor, target, target);
+    }
+
+    private Boolean lookupPlugin(String name) {
+        return Option.of(Jenkins.getInstance().getPlugin(name)).isDefined();
+    }
+
+    private HashMap<String, Descriptor<T>> handleDuplicateSymbols(HashMap<String, Descriptor<T>> r, Tuple2<String, Descriptor<T>> t) {
```

Comment:
it's not very clear what the parameters are here, what are r and t?

also do we need to be using HashMap or can we use Map?